### DISCUSSION TITLE
[Idea] Bumping Doctrine DBAL to version 3 in PrestaShop 9, having multiple branches for impacted modules

### DISCUSSION THREAD

**Discussion by matks** [PrestaShop/PrestaShop#33511] [Ideas]
### The problem

Hello folks,

So @M0rgan01 was trying to update [Composer dependency [egulias/email-validator to 4.0.1 for PrestaShop 9](https://github.com/PrestaShop/PrestaShop/pull/33459).

But while doing that he figured out he also needed to update... doctrine/dbal . And other doctrine packages.

Bumping Doctrine to a major version does bring some Breaking Changes for both the Core and modules.
Don't forget modules can use Doctrine entities so this will impact developer modules too.

While analyzing the topic @M0rgan01 and I figured out that we will need to update some modules that do rely on Doctrine entities. The problem is that this update will be a not backward compatible update: the module, once updated, will not be compatible with PrestaShop 8 or PrestaShop 1.7

**What does this mean? Let's take the example of productcomments**

Current productcomments version is 5.0.3

If we update productcomments to be compatible with doctrine we need to release a version 6.0.0 of productcomments.

However this version 6.0.0 will not be compatible with PS 1.7, PS 8.0 and PS 8.1 . productcomments 6.0.0 minimum PrestaShop version will be 9.0.0 . And PrestaShop 9.0.0 release date is set to be somewhere between end of 2023 and beginning of 2024.

We cannot simply release productcomments 6.0.0 with PrestaShop minimum version to 9.0.0 and tell to the prestashop community "Hey, if you want to submit PRs for productcomments module, you can but they will be released with version 6 so you wont be able to use it on a live stable version of PrestaShop before 2024" 😭 

### The solution

So... we probably need to create 2 branches for productcomments (and any other impacted module):
- `dev` branch would become version 6.0.0, with PrestaShop minimum version to 9.0.0
- `5.x` branch would be compatible with PS 8 and be used to release 5.x minor and patch versions

Which means branch merge work to do, more maintainance, more time to spend just to keep the system running 😭 

### Do you plan to work on this subject?

- [X] I'm willing to contribute a [formal specification](https://github.com/PrestaShop/prestashop-specs).
- [X] I'm willing to provide any wireframes or design assets required for this feature.
- [X] I'm willing to submit a Pull Request that implements this feature.
- [X] I'm willing to help verify that the implemented feature works as intended and produces no unintended side effects.

**Comment by nicosomb** [comment]
@matks Thank you for this post.

Can you list all the impacted modules?

**Comment by kpodemski** [comment]
We knew it would happen at some point. I think the proposed solution is valid, the thing is that we have **a lot** of modules. We need to have a tool to handle merges better. Maybe we could track it, we could think of improving the Module Dashboard :-) cc @Hlavtox 

https://build.prestashop-project.org/ps-monitor-module-releases/

**Comment by leemyongpakvn** [comment]
We have done a lot of Breaking Changes in PS 9, and fortunately we decided to keep them isolated inside PS core till now ;)

**Comment by eternoendless** [comment]
If the module doesn't include the doctrine dependencies itself, what is the actual impact for modules using doctrine entities?

**Reply by M0rgan01** [reply]
If I take as an example, the module productcomments, the changes are present in the repository part, there are: 

-  changes to avoid using deprecated functions, ex: `$qb->execute()->fetchAll()` -> `$qb->executeQuery()->fetchAllAssociative()`
-  mandatory changes because breaking change, ex: `$qb->execute()->fetchColumn()` -> `$qb->execute()->fetchOne()`

Here is an example PR https://github.com/PrestaShop/productcomments/pull/178

**Reply by kpodemski** [reply]
not the cleanest solution but you could check the PS version inside this methods, or use adapter pattern to handle few versions at once

**Reply by M0rgan01** [reply]
I thought about that too, but I don't find it very clean. There are already pieces of code that are specific to the version of prestashop core. Like `if (version_compare(_PS_VERSION_, '1.7.5', '>='))...`, but this is for very specific, very restricted pieces of code. In addition, it is necessary to manage PHPstan errors according to the version of Prestashop

**Reply by leemyongpakvn** [reply]
productcomments is based on Core's doctrine  as declared at it's [common.yml](https://github.com/PrestaShop/productcomments/blob/dev/config/common.yml)
In fact, DoctrineDBAL 3.6.5 is backward-compatible with DoctrineDBAL 2.7 as coded at its [Result](https://github.com/doctrine/dbal/blob/3.6.5/src/Result.php) and [Statement](https://github.com/doctrine/dbal/blob/3.6.5/src/Statement.php), so update DBAL in PS Core to v3.6.5 won't break productcomments module.
If we apply above PR 178, we really break backward-compatibility of this module.

**Reply by M0rgan01** [reply]
> update DBAL in PS Core to v3.6.5 won't break productcomments module.

Here is the screenshot which is the result of my UI test in error on the CI. https://github.com/PrestaShop/PrestaShop/pull/33459
![fail_test_1](https://github.com/PrestaShop/PrestaShop/assets/36233446/36759378-0ca1-48de-a627-361b2dc90d5c)

I would have also hoped that there is no error of this type ^^

**Reply by leemyongpakvn** [reply]
@M0rgan01 Yeah, as coded at DBAL 3.6.5 https://github.com/doctrine/dbal/blob/3.6.5/src/Result.php#L270-L289, you can use
`fetch(FetchMode::COLUMN)` instead of `fetchColumn`

**Reply by M0rgan01** [reply]
Might as well go to fetchOne, which is not deprecated :) There will be the same problem below, because now the execute method returns a Result object, which contains the fetch function.

![image](https://github.com/PrestaShop/PrestaShop/assets/36233446/87d4cebb-1603-4414-8e84-74a3fa519a79)

![image](https://github.com/PrestaShop/PrestaShop/assets/36233446/2f35cdfb-2542-4953-894b-b5c88f818836)

**Reply by leemyongpakvn** [reply]
Good news ;) We can use DBAL connection executeQuery instead , it is available in both [DBAL 3.6.5](https://github.com/doctrine/dbal/blob/3.6.5/src/Connection.php#L1069) and [DBAL 2.7.2](https://github.com/doctrine/dbal/blob/v2.7.2/lib/Doctrine/DBAL/Connection.php#L910). You can see an example at https://github.com/PrestaShop/productcomments/pull/177/files#diff-5601ffa051870b372f5efe0c2e5ba6dad8cab3dabedac62ac36a31901f6396e7R279-R282

**Reply by M0rgan01** [reply]
We might be able to get away with the module with that. We can also use Doctrine\DBAL\Exception which already exists, instead of DBALException which disappears. After a few changes it should be fine :) Thanks
